### PR TITLE
Color map rescale refactor

### PR DESCRIPTION
**Pull request by brigb123**
ColorMap.cc object is no longer in SRInterface.cc, it is only seen from CreateStandardColorMap.cc -> ShowField.cc modules. 

All algorithms for retrieving color map values are contained in ColorMap.cc and taken out of shaders.

Did regression tests with:

1. rescaleColorMap.srn5
2. test_basis.scn5
3. simple latVol
4. simple colorMap
5. tikinov
6. inverseForwardProblem
7. simple boundingBox

### CODE REVIEW

File: src/Core/Datatypes/ColorMap.h [hunk 1]
```diff
@@ -53,10 +54,11 @@ namespace Datatypes {
     bool getColorMapInvert() const;
     double getColorMapRescaleScale() const;
     double getColorMapRescaleShift() const;
-    double getColorMapActualMin() const;
-    double getColorMapActualMax() const;
-    Core::Datatypes::ColorRGB getColorMapVal(float v) const;
-    float getTransformedColor(float v) const;
+    Core::Datatypes::ColorRGB getColorMapVal(double v) const;
+    Core::Datatypes::ColorRGB valueToColor(double scalar);
+    Core::Datatypes::ColorRGB valueToColor(Core::Geometry::Tensor tensor);
```
Comment (dcwhite) [inline src/Core/Datatypes/ColorMap.h]: @brigb123 pass by const &

File: src/Core/Datatypes/ColorMap.h [hunk 2]
```diff
@@ -53,10 +54,11 @@ namespace Datatypes {
...
+    Core::Datatypes::ColorRGB valueToColor(Core::Geometry::Vector vector);
```
Comment (dcwhite) [inline src/Core/Datatypes/ColorMap.h]: @brigb123 pass by const &

File: src/Interface/Modules/Render/ES/SRInterface.cc [hunk 3]
```diff
@@ -582,46 +578,8 @@ namespace SCIRun {
...
-          //mCore.displayEntityVersusSystemInfo(entityID, getSystemName_RenderColorMap());
-        }
-        else
-        {
-          std::cerr << "Renderer: Unknown color scheme!" << std::endl;
-          RenderBasicGeom geom;
-          mCore.addComponent(entityID, geom);
-        }
+        RenderBasicGeom geom;
+        mCore.addComponent(entityID, geom);
```
Comment (dcwhite) [inline src/Interface/Modules/Render/ES/SRInterface.cc]: So much deletion = :)

File: src/Modules/Visualization/ShowField.cc [hunk 4]
```diff
@@ -240,81 +241,45 @@ GeometryHandle ShowFieldModule::buildGeometryObject(
   const int dim = field->vmesh()->dimensionality();
   if (showEdges && dim < 1) { showEdges = false; }
   if (showFaces && dim < 2) { showFaces = false; }
+  
+  //adjust color map to scalar data if there is one.
+  ColorMap *cmap = colorMap.get().get();
+  ColorMap adjustedCMap(cmap->getColorMapName(),cmap->getColorMapResolution(),
+                        cmap->getColorMapShift(), cmap->getColorMapInvert(),
+                        cmap->getColorMapRescaleScale(), cmap->getColorMapRescaleShift());
+  if (colorMap) {
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: On line 246 you've already dereferenced the colorMap pointer, so this check should happen earlier.

File: src/Modules/Visualization/ShowField.cc [hunk 5]
```diff
@@ -240,81 +241,45 @@ GeometryHandle ShowFieldModule::buildGeometryObject(
...
-  //   float biggest = Max(Abs(minmax_.first), Abs(minmax_.second));
-  //   minmax_.first  = -biggest;
-  //   minmax_.second =  biggest;
-  // }
-}
 
 void ShowFieldModule::renderFaces(
   boost::shared_ptr<SCIRun::Field> field,
-  boost::optional<boost::shared_ptr<SCIRun::Core::Datatypes::ColorMap>> colorMap,
+  ColorMap colorMap,
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: pass by const &

File: src/Modules/Visualization/ShowField.cc [hunk 6]
```diff
@@ -344,7 +309,7 @@ void ShowFieldModule::renderFaces(
 
 void ShowFieldModule::renderFacesLinear(
   boost::shared_ptr<SCIRun::Field> field,
-  boost::optional<boost::shared_ptr<SCIRun::Core::Datatypes::ColorMap>> colorMap,
+  ColorMap colorMap,
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: pass by const &

File: src/Modules/Visualization/ShowField.cc [hunk 7]
```diff
@@ -361,8 +326,8 @@ void ShowFieldModule::renderFacesLinear(
   std::vector<Core::Geometry::Vector> vvals(10);
   std::vector<Core::Geometry::Tensor> tvals(10);
 
-  std::vector<Material> vcols(10, Material());
-  std::vector<double> scols(10);
+  std::vector<ColorRGB> face_colors;
+  face_colors.resize(10);
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: Why 10? Use a named constant, no magic numbers.

File: src/Modules/Visualization/ShowField.cc [hunk 8]
```diff
@@ -409,8 +365,8 @@ void ShowFieldModule::renderFacesLinear(
   // Attempt some form of precalculation of iboBuffer and vboBuffer size.
   // This Initial size estimation will be off quite a bit. Each face will
   // have up to 3 nodes associated.
-  uint32_t iboSize = mesh->num_faces() * sizeof(uint32_t) * 3;
-  uint32_t vboSize = mesh->num_faces() * sizeof(float) * 3;
+  uint32_t iboSize = static_cast<uint32_t>(mesh->num_faces() * sizeof(uint32_t) * 3);
+  uint32_t vboSize = static_cast<uint32_t>(mesh->num_faces() * sizeof(float) * 7);
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: Same here, name the constants 3 and 7. The 7 was a 3, reader has no idea why it changed.

File: src/Modules/Visualization/ShowField.cc [hunk 9]
```diff
@@ -646,17 +600,9 @@ void ShowFieldModule::renderFacesLinear(
...
-    uniforms.push_back(GeometryObject::SpireSubPass::Uniform("uCMInvert",map->getColorMapInvert()?1.f:0.f));
-    uniforms.push_back(GeometryObject::SpireSubPass::Uniform("uCMShift",static_cast<float>(map->getColorMapShift())));
-    uniforms.push_back(GeometryObject::SpireSubPass::Uniform("uCMResolution",static_cast<float>(map->getColorMapResolution())));
-    double scl = map->getColorMapRescaleScale(), shft = map->getColorMapRescaleShift();
-    uniforms.push_back(GeometryObject::SpireSubPass::Uniform("uRescaleScale",static_cast<float>(scl)));
-    uniforms.push_back(GeometryObject::SpireSubPass::Uniform("uRescaleShift",static_cast<float>(shft)));
+    attribs.push_back(GeometryObject::SpireVBO::AttributeData("aColor", 4 * sizeof(float)));
 
-    if (state.get(RenderState::IS_DOUBLE_SIDED) == false)
+    if (!state.get(RenderState::IS_DOUBLE_SIDED))
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: Thanks for this change, I despise the == falses.

File: src/Modules/Visualization/ShowField.cc [hunk 10]
```diff
@@ -831,8 +770,7 @@ void ShowFieldModule::addFaceGeom(
   CPM_VAR_BUFFER_NS::VarBuffer* iboBuffer,
   CPM_VAR_BUFFER_NS::VarBuffer* vboBuffer,
   GeometryObject::ColorScheme colorScheme,
-  std::vector<double> &scols,
-  std::vector<Material> &vcols,
+  std::vector<ColorRGB> &face_colors,
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: If possible, pass by const &

File: src/Modules/Visualization/ShowField.cc [hunk 11]
```diff
@@ -1221,17 +1153,10 @@ void ShowFieldModule::renderNodes(
...
   if (colorScheme == GeometryObject::COLOR_MAP) {
-    attribs.push_back(GeometryObject::SpireVBO::AttributeData("aFieldData", 1 * sizeof(float)));
-    //push the color map parameters
-    ColorMap * map = colorMap.get().get();
-    uniforms.push_back(GeometryObject::SpireSubPass::Uniform("uCMInvert",map->getColorMapInvert()?1.f:0.f));
-    uniforms.push_back(GeometryObject::SpireSubPass::Uniform("uCMShift",static_cast<float>(map->getColorMapShift())));
-    uniforms.push_back(GeometryObject::SpireSubPass::Uniform("uCMResolution",static_cast<float>(map->getColorMapResolution())));
-    uniforms.push_back(GeometryObject::SpireSubPass::Uniform("uRescaleScale",static_cast<float>(map->getColorMapRescaleScale())));
-    uniforms.push_back(GeometryObject::SpireSubPass::Uniform("uRescaleShift",static_cast<float>(map->getColorMapRescaleShift())));
+    attribs.push_back(GeometryObject::SpireVBO::AttributeData("aColor", 4 * sizeof(float)));
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: Another magic number

File: src/Modules/Visualization/ShowField.cc [hunk 12]
```diff
@@ -1406,28 +1326,21 @@ void ShowFieldModule::renderNodes(
...
-  }
-
   geom->mPasses.push_back(pass);
 }
 
 
 void ShowFieldModule::renderEdges(
   boost::shared_ptr<SCIRun::Field> field,
-  boost::optional<boost::shared_ptr<SCIRun::Core::Datatypes::ColorMap>> colorMap,
+  ColorMap colorMap,
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: pass by const &

File: src/Modules/Visualization/ShowField.cc [hunk 13]
```diff
@@ -1652,24 +1553,21 @@ void ShowFieldModule::renderEdges(
       double theta_inc = 2. * M_PI / num_strips, phi_inc = M_PI / num_strips;
       std::vector<Point> epts = { { p0, p1 } };
       for (auto a : epts) {
-        double col = a == p0 ? scol0 : scol1;
-        Material rgbcol = a == p0 ? vcol0 : vcol1;
+        ColorRGB col = a == p0 ? edge_colors[0] : edge_colors[1];
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: Can use for (const auto& a : ...)

Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: Or even better:
```
for (const auto& a : { { p0, p1 } } )
```

File: src/Modules/Visualization/ShowField.h [hunk 14]
```diff
@@ -105,20 +106,20 @@ namespace SCIRun {
         /// @{
         void renderNodes(
           boost::shared_ptr<SCIRun::Field> field,
-          boost::optional<boost::shared_ptr<SCIRun::Core::Datatypes::ColorMap>> colorMap,
+          SCIRun::Core::Datatypes::ColorMap colorMap,
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.h]: pass by const & (etc below, already noted for .cc)

**Comment by dcwhite** [comment]
Face transparency broken on Windows now. Compared to master, where it's not broken. I'll check Mac

**Comment by dcwhite** [comment]
thanks

File: src/Modules/Visualization/ShowField.cc [hunk 15]
```diff
@@ -240,81 +241,45 @@ GeometryHandle ShowFieldModule::buildGeometryObject(
   const int dim = field->vmesh()->dimensionality();
   if (showEdges && dim < 1) { showEdges = false; }
   if (showFaces && dim < 2) { showFaces = false; }
+  //adjust color map to scalar data if there is one.
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: @brigb123 Finally, is this PR supposed to completely implement the RescaleColorMap module? If so, this block should just be removed--ShowField does not do any min/max scaling. If you intend to remove this section separately, let me know.

File: src/Modules/Visualization/ShowField.cc [hunk 16]
```diff
@@ -240,81 +241,45 @@ GeometryHandle ShowFieldModule::buildGeometryObject(
...
+    adjustedCMap = ColorMap(cmap->getColorMapName(),cmap->getColorMapResolution(),
+                          cmap->getColorMapShift(), cmap->getColorMapInvert(),
+                          cmap->getColorMapRescaleScale(), cmap->getColorMapRescaleShift());
+    double actual_max = std::numeric_limits<double>::min();
+    double actual_min = std::numeric_limits<double>::max();
+    VField * fld = field->vfield();
+    if (fld->minmax(actual_min, actual_max)) {
+        adjustedCMap.adjustColorMapForData(actual_min, actual_max);
+    }
+  }
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowField.cc]: //end section to be removed

File: src/Modules/Visualization/ShowField.cc (hunk 15 above)
Comment (brigb123) [inline src/Modules/Visualization/ShowField.cc]: I finally get it... yes, i'll take it out and check that it works as intended with/without rescale.

**Pull request merged**